velox/functions/sparksql/DateTimeFunctions.cpp

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/DateTimeFunctions.h"
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+std::optional<Timestamp> makeTimeStampFromDecodedArgs(
+    vector_size_t row,
+    DecodedVector* yearVector,
+    DecodedVector* monthVector,
+    DecodedVector* dayVector,
+    DecodedVector* hourVector,
+    DecodedVector* minuteVector,
+    DecodedVector* microsVector) {
+  // Check hour.
+  auto hour = hourVector->valueAt<int32_t>(row);
+  if (hour < 0 || hour > 24) {
+    return std::nullopt;
+  }
+  // Check miniute.
+  auto minute = minuteVector->valueAt<int32_t>(row);
+  if (minute < 0 || minute > 60) {
+    return std::nullopt;
+  }
+  // Check microseconds.
+  auto micros = microsVector->valueAt<int64_t>(row);
+  if (micros < 0) {
+    return std::nullopt;
+  }
+  auto seconds = micros / util::kMicrosPerSec;
+  if (seconds > 60 || (seconds == 60 && micros % util::kMicrosPerSec != 0)) {
+    return std::nullopt;
+  }
+
+  // year, month, day will be checked in utils::daysSinceEpochFromDate;
+  try {
+    auto daysSinceEpoch = util::daysSinceEpochFromDate(
+        yearVector->valueAt<int32_t>(row),
+        monthVector->valueAt<int32_t>(row),
+        dayVector->valueAt<int32_t>(row));
+    auto localMicros =
+        hour * util::kMicrosPerHour + minute * util::kMicrosPerMinute + micros;
+    return util::fromDatetime(daysSinceEpoch, localMicros);
+  } catch (const VeloxException& e) {
```

Comment:
If user error can only be `VeloxUserError` here, maybe we can just catch such exception.